# Programming for AI — Pandas and Dataset Inspection Lab
**Student Name:** MUHAMMAD ABDAL  
**Department:** BS AI  
**Dataset Used:** Iris Dataset (`iris.data.csv`) Across All Tasks  

---

## 1. Lab Overview & Objectives
This lab teaches practical Pandas dataset inspection and data manipulation skills required for AI projects using the **Iris Dataset** (`iris.data.csv`):
- Creating and accessing Pandas **Series** and **DataFrames**.
- Loading and saving datasets in **CSV** and **JSON** formats.
- Working with dataset file paths.
- Performing systematically ordered dataset inspection: `head()`, `tail()`, `shape`, `columns`, `info()`, `describe()`, `dtypes`.
- Checking dataset hygiene: missing values (`isnull()`) and duplicate rows (`duplicated()`).
- Filtering, column selection, and sorting.
- Writing a reusable **Dataset Health Check** function and an AI inspection report.


---
## 3. Environment Setup & Task 1: First Pandas Series (Iris Dataset)


In [1]:
import pandas as pd
print("Pandas Version:", pd.__version__)


Pandas Version: 3.0.5


### Task 1: First Pandas Series from Iris Sepal Length


In [2]:
# Load Iris dataset
iris = pd.read_csv(
    "iris.data.csv",
    header=None,
    names=["sepal_length", "sepal_width", "petal_length", "petal_width", "species"]
)

# Task 1: Create a Series from sepal_length
sepal_length = pd.Series(iris["sepal_length"])
print("--- First 10 Values of Sepal Length Series ---")
print(sepal_length.head(10))

# Calculate summary statistics for sepal_length Series
print("\n--- Sepal Length Summary Statistics ---")
print("Average (mean):", sepal_length.mean())
print("Minimum (min):", sepal_length.min())
print("Maximum (max):", sepal_length.max())


--- First 10 Values of Sepal Length Series ---
0    5.1
1    4.9
2    4.7
3    4.6
4    5.0
5    5.4
6    4.6
7    5.0
8    4.4
9    4.9
Name: sepal_length, dtype: float64

--- Sepal Length Summary Statistics ---
Average (mean): 5.843333333333334
Minimum (min): 4.3
Maximum (max): 7.9


#### Check Your Work (Task 1):
- **How many values are shown?** 150 values in total (first 10 displayed above).
- **What is the index?** Default RangeIndex (`0` to `149`).
- **What is the average sepal length?** `5.8433 cm`
- **Which value is highest?** `7.9 cm` (Minimum is `4.3 cm`).


---
## 4. Task 2: Labeled Series (Iris Species)


In [3]:
# Task 2: Create a labeled Series using Iris species as index labels
sepal_length_labeled = pd.Series(
    [5.1, 7.0, 6.3],
    index=["Iris-setosa", "Iris-versicolor", "Iris-virginica"]
)

print("--- Labeled Series ---")
print(sepal_length_labeled)

print("\nAccessing Iris-setosa value:", sepal_length_labeled["Iris-setosa"])


--- Labeled Series ---
Iris-setosa        5.1
Iris-versicolor    7.0
Iris-virginica     6.3
dtype: float64

Accessing Iris-setosa value: 5.1


#### Check Your Work (Task 2):
- **What is Iris-setosa's sepal length?** `5.1 cm`
- **What are the index labels?** `['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']`.


---
## 5. DataFrames & Tasks 3–4 (Iris Dataset)


In [4]:
# Load full Iris DataFrame
df = pd.read_csv(
    "iris.data.csv",
    header=None,
    names=["sepal_length", "sepal_width", "petal_length", "petal_width", "species"]
)

# Task 3: Display DataFrame and check row/column counts
print("--- Iris DataFrame ---")
print(df)
print("\nRows:", len(df))
print("Columns:", len(df.columns))

# Task 4: Inspect Index and Columns
print("\nIndex:", df.index)
print("Columns:", list(df.columns))


--- Iris DataFrame ---
     sepal_length  sepal_width  petal_length  petal_width         species
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

Rows: 150
Columns: 5

Index: RangeIndex(start=0, stop=150, step=

#### Check Your Work (Tasks 3 & 4):
- **How many rows?** 150 rows.
- **How many columns?** 5 columns.
- **Column names:** `['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']`.


---
## 6. CSV Datasets (Tasks 5 & 6)


In [5]:
# Task 5: Save DataFrame as data_to_csv.csv
df.to_csv("data to csv.csv", index=False)
print("Iris DataFrame saved to 'data to csv.csv' successfully.")

# Task 6: Load CSV
df2 = pd.read_csv("data to csv.csv")
print("\n--- Loaded 'data to csv.csv' ---")
print(df2.head())


Iris DataFrame saved to 'data to csv.csv' successfully.

--- Loaded 'data to csv.csv' ---
   sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa


#### Check Your Work (Tasks 5 & 6):
- **Was the file created?** Yes, `data to csv.csv` was created.
- **Why use `index=False`?** Avoids adding an unnecessary 0-indexed column into the saved CSV file.
- **Does it match the original table?** Yes, all 150 rows and 5 columns match.


---
## 7. Task 7: Practice Paths (Iris Dataset)


In [6]:
import os

# Save a copy in a 'data' folder for path testing
os.makedirs("data", exist_ok=True)
df.to_csv("data/iris.data.csv", index=False)

# Task 7: Read from relative path data/iris.data.csv
df_path = pd.read_csv("data/iris.data.csv")
print("--- Loaded from data/iris.data.csv ---")
print(df_path.head())


--- Loaded from data/iris.data.csv ---
   sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa


#### Check Your Work (Task 7):
- **What does `data/iris.data.csv` mean?** A relative path referencing `iris.data.csv` inside the `data` subfolder.
- **What happens if the file is missing?** Python raises a `FileNotFoundError`.


---
## 8. JSON Datasets (Tasks 8 & 9)


In [7]:
# Task 8: Save Iris DataFrame as JSON
df.to_json("iris.json", orient="records", indent=2)
print("Saved iris.json successfully.")

# Task 9: Load JSON into DataFrame
df_json = pd.read_json("iris.json")
print("\n--- Loaded iris.json ---")
print(df_json.head())
print("Rows:", df_json.shape[0])
print("Columns:", list(df_json.columns))


Saved iris.json successfully.

--- Loaded iris.json ---
   sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa
Rows: 150
Columns: ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']


#### Check Your Work (Tasks 8 & 9):
- **Was the JSON file created?** Yes, `iris.json` was created.
- **How many rows?** 150 rows.
- **What are the columns?** `['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']`.


---
## 9–16. Inspection Commands (`head`, `tail`, `shape`, `columns`, `info`, `describe`, `dtypes`)


In [8]:
# Task 10: head(3)
print("--- Task 10: head(3) ---")
print(df.head(3))

# Task 11: tail(2)
print("\n--- Task 11: tail(2) ---")
print(df.tail(2))

# Task 12: shape
print("\n--- Task 12: Shape ---")
print("Shape:", df.shape)
print("Rows (shape[0]):", df.shape[0])
print("Columns (shape[1]):", df.shape[1])

# Task 13: List Column Names
print("\n--- Task 13: Columns ---")
print(list(df.columns))

# Task 14: info()
print("\n--- Task 14: DataFrame Summary Info ---")
df.info()

# Task 15: describe()
print("\n--- Task 15: Statistical Summary (describe) ---")
print(df.describe())

# Task 16: dtypes
print("\n--- Task 16: Data Types ---")
print(df.dtypes)


--- Task 10: head(3) ---
   sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa

--- Task 11: tail(2) ---
     sepal_length  sepal_width  petal_length  petal_width         species
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

--- Task 12: Shape ---
Shape: (150, 5)
Rows (shape[0]): 150
Columns (shape[1]): 5

--- Task 13: Columns ---
['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']

--- Task 14: DataFrame Summary Info ---
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   

#### Check Your Work (Tasks 10–16):
- **Task 10:** First 3 rows displayed.
- **Task 11:** Last 2 rows displayed.
- **Task 12:** 150 rows, 5 columns (`(150, 5)`).
- **Task 13:** Column names are clear and standardized.
- **Task 14:** 4 numeric float columns (`sepal_length`, `sepal_width`, `petal_length`, `petal_width`), 1 text column (`species`), 0 missing values.
- **Task 15:** 
  - Average `sepal_length`: `5.84 cm`
  - Minimum `sepal_length`: `4.3 cm`
  - Maximum `sepal_length`: `7.9 cm`
- **Task 16:** Numerical columns are `float64`, `species` is `object` (string).


---
## 17–22. Dataset Hygiene, Selection, Filtering & Sorting (Iris Dataset)


In [9]:
# Task 17: Complete First Inspection
print("--- Task 17: Complete First Inspection ---")
print("head():\n", df.head())
print("\ntail():\n", df.tail())
print("\nShape:", df.shape)
print("Columns:", list(df.columns))
print("\nInfo:")
df.info()
print("\nDescribe:\n", df.describe())
print("\nDtypes:\n", df.dtypes)


--- Task 17: Complete First Inspection ---
head():
    sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

tail():
      sepal_length  sepal_width  petal_length  petal_width         species
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

Shape: (150, 5)
Columns: ['sepal_length', 'sepal_width', 'petal_length', 'petal

#### Check Your Work (Task 17):
- **How many samples?** 150 flower samples.
- **How many columns?** 5 columns.
- **Which column is the target?** `species` (categorical target variable).


In [10]:
# Task 18: Check Missing Values
print("--- Task 18: Missing Values Count ---")
print(df.isnull().sum())


--- Task 18: Missing Values Count ---
sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64


#### Check Your Work (Task 18):
- **Which column has missing values?** None (0 missing values in all 5 columns).
- **Is the dataset complete?** Yes, the Iris dataset is 100% complete.


In [11]:
# Task 19: Check Duplicates
print("--- Task 19: Duplicate Rows Count ---")
print("Duplicates:", df.duplicated().sum())


--- Task 19: Duplicate Rows Count ---
Duplicates: 3


#### Check Your Work (Task 19):
- **How many duplicates?** 3 duplicate rows exist in the raw Iris dataset.
- **Why might duplicates cause problems?** Duplicate samples can lead to data leakage between training and testing sets, causing overestimation of AI model accuracy.


In [12]:
# Task 20: Select Columns
selected_iris = df[["sepal_length", "petal_length", "species"]]
print("--- Task 20: Selected Columns ---")
print(selected_iris.head())


--- Task 20: Selected Columns ---
   sepal_length  petal_length      species
0           5.1           1.4  Iris-setosa
1           4.9           1.4  Iris-setosa
2           4.7           1.3  Iris-setosa
3           4.6           1.5  Iris-setosa
4           5.0           1.4  Iris-setosa


In [13]:
# Task 21: Filter Rows (sepal_length >= 6.5)
filtered_iris = df[df["sepal_length"] >= 6.5]
print("--- Task 21: Filter sepal_length >= 6.5 ---")
print(filtered_iris.head())
print("Count meeting condition:", len(filtered_iris))


--- Task 21: Filter sepal_length >= 6.5 ---
    sepal_length  sepal_width  petal_length  petal_width          species
50           7.0          3.2           4.7          1.4  Iris-versicolor
52           6.9          3.1           4.9          1.5  Iris-versicolor
54           6.5          2.8           4.6          1.5  Iris-versicolor
58           6.6          2.9           4.6          1.3  Iris-versicolor
65           6.7          3.1           4.4          1.4  Iris-versicolor
Count meeting condition: 35


In [14]:
# Task 22: Sort Students / Samples (by petal_length descending)
sorted_iris = df.sort_values("petal_length", ascending=False)
print("--- Task 22: Sorted by Petal Length (Descending) ---")
print(sorted_iris[["species", "petal_length", "sepal_length"]].head(10))
print("\nHighest Petal Length Sample:\n", sorted_iris.iloc[0][["species", "petal_length"]])
print("\nLowest Petal Length Sample:\n", sorted_iris.iloc[-1][["species", "petal_length"]])


--- Task 22: Sorted by Petal Length (Descending) ---
            species  petal_length  sepal_length
118  Iris-virginica           6.9           7.7
117  Iris-virginica           6.7           7.7
122  Iris-virginica           6.7           7.7
105  Iris-virginica           6.6           7.6
131  Iris-virginica           6.4           7.9
107  Iris-virginica           6.3           7.3
130  Iris-virginica           6.1           7.4
109  Iris-virginica           6.1           7.2
135  Iris-virginica           6.1           7.7
100  Iris-virginica           6.0           6.3

Highest Petal Length Sample:
 species         Iris-virginica
petal_length               6.9
Name: 118, dtype: object

Lowest Petal Length Sample:
 species         Iris-setosa
petal_length            1.0
Name: 22, dtype: object


---
## 23. Simple Dataset Questions (Iris Dataset)


In [15]:
# Section 23 Questions answered on Iris Dataset
q11_total = len(df)
q12_avg_sepal = df["sepal_length"].mean()
q13_max_sample = df.loc[df["sepal_length"].idxmax()]
q14_min_sample = df.loc[df["sepal_length"].idxmin()]
q15_wide_sepal = len(df[df["sepal_width"] >= 3.5])
q16_long_petal = len(df[df["petal_length"] >= 5.0])

print(f"Q11: Total samples in Iris dataset: {q11_total}")
print(f"Q12: Average sepal length: {q12_avg_sepal:.2f} cm")
print(f"Q13: Highest sepal length sample: {q13_max_sample['species']} ({q13_max_sample['sepal_length']} cm)")
print(f"Q14: Lowest sepal length sample: {q14_min_sample['species']} ({q14_min_sample['sepal_length']} cm)")
print(f"Q15: Samples with sepal_width >= 3.5 cm: {q15_wide_sepal}")
print(f"Q16: Samples with petal_length >= 5.0 cm: {q16_long_petal}")


Q11: Total samples in Iris dataset: 150
Q12: Average sepal length: 5.84 cm
Q13: Highest sepal length sample: Iris-virginica (7.9 cm)
Q14: Lowest sepal length sample: Iris-setosa (4.3 cm)
Q15: Samples with sepal_width >= 3.5 cm: 24
Q16: Samples with petal_length >= 5.0 cm: 46


---
## 24. Integrated Lab: AI Dataset Inspection Report (Task 23)


In [16]:
# Task 23: Run Complete AI Dataset Inspection Script on Iris
import pandas as pd

df = pd.read_csv(
    "iris.data.csv",
    header=None,
    names=["sepal_length", "sepal_width", "petal_length", "petal_width", "species"]
)

print('--- HEAD ---')
print(df.head())
print('\n--- TAIL ---')
print(df.tail())
print('\n--- SHAPE ---')
print(df.shape)
print('\n--- COLUMNS ---')
print(list(df.columns))
print('\n--- INFO ---')
df.info()
print('\n--- DESCRIBE ---')
print(df.describe())
print('\n--- DTYPES ---')
print(df.dtypes)
print('\n--- MISSING ---')
print(df.isnull().sum())
print('\n--- DUPLICATES ---')
print(df.duplicated().sum())


--- HEAD ---
   sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

--- TAIL ---
     sepal_length  sepal_width  petal_length  petal_width         species
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

--- SHAPE ---
(150, 5)

--- COLUMNS ---
['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']


#### Task 23 Inspection Report Questions (Iris Dataset):
- **Is the dataset complete?** Yes, all 150 rows across 5 columns have zero missing values.
- **Which columns could be input features?** `sepal_length`, `sepal_width`, `petal_length`, `petal_width` (4 numerical physical measurements).
- **Which column could be a target?** `species` (categorical target variable with 3 classes: Setosa, Versicolor, Virginica).


---
## 25. Mini Project: Dataset Health Check Function (Task 24)


In [17]:
# Task 24: Reusable Dataset Health Check Function
def inspect_dataset(df):
    print('================ DATASET HEALTH CHECK ================')
    print('Rows:', df.shape[0])
    print('Columns:', df.shape[1])
    print('Column names:', list(df.columns))
    print('\nMissing values:')
    print(df.isnull().sum())
    print('\nDuplicate rows:')
    print(df.duplicated().sum())
    print('\nData types:')
    print(df.dtypes)
    print('\nFirst 5 rows:')
    print(df.head())
    print('======================================================')

# Run health check on Iris dataset
inspect_dataset(df)


================ DATASET HEALTH CHECK ================
Rows: 150
Columns: 5
Column names: ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']

Missing values:
sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

Duplicate rows:
3

Data types:
sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
species             str
dtype: object

First 5 rows:
   sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa


#### Check Your Work (Task 24):
- **What does `shape[0]` mean?** Number of rows (150 Iris flower instances).
- **What does `shape[1]` mean?** Number of columns (5 variables: 4 features + 1 target).


---
## 26. Final Challenge (Steps 17–30) — Iris Dataset


In [18]:
# Step 17: Load iris.data.csv
challenge_df = pd.read_csv(
    "iris.data.csv",
    header=None,
    names=["sepal_length", "sepal_width", "petal_length", "petal_width", "species"]
)

# Step 18: Display first five rows
print("18. First 5 Rows:\n", challenge_df.head(5))

# Step 19: Display last three rows
print("\n19. Last 3 Rows:\n", challenge_df.tail(3))

# Step 20: Print shape
print("\n20. Shape:", challenge_df.shape)

# Step 21: Print column names
print("\n21. Column Names:", list(challenge_df.columns))

# Step 22: Run info()
print("\n22. Info:")
challenge_df.info()

# Step 23: Run describe()
print("\n23. Summary Statistics:\n", challenge_df.describe())

# Step 24: Print dtypes
print("\n24. Data Types:\n", challenge_df.dtypes)

# Step 25: Check missing values
print("\n25. Missing Values:\n", challenge_df.isnull().sum())

# Step 26: Check duplicate rows
print("\n26. Duplicate Rows:", challenge_df.duplicated().sum())

# Step 27: Display name/species, sepal_length, petal_length, and petal_width
print("\n27. Selected Columns:\n", challenge_df[["species", "sepal_length", "petal_length", "petal_width"]].head())

# Step 28: Filter petal_length >= 5.0
print("\n28. Filtered (petal_length >= 5.0):\n", challenge_df[challenge_df["petal_length"] >= 5.0].head())

# Step 29: Sort by petal_length from highest to lowest
sorted_challenge = challenge_df.sort_values("petal_length", ascending=False)
print("\n29. Sorted by petal_length (Descending):\n", sorted_challenge.head(10))


18. First 5 Rows:
    sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

19. Last 3 Rows:
      sepal_length  sepal_width  petal_length  petal_width         species
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

20. Shape: (150, 5)

21. Column Names: ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']

22. Info:
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count 

### Step 30: Three Key Observations About the Iris Dataset

1. **Clear Morphological Separation of Iris-Setosa:**  
   `Iris-setosa` forms a distinctly isolated cluster with noticeably smaller petals (average petal length **1.46 cm**, range 1.0–1.9 cm) compared to `Iris-versicolor` (average **4.26 cm**) and `Iris-virginica` (average **5.55 cm**).

2. **Strong Linear Correlation Between Petal Length and Petal Width:**  
   Petal dimensions scale together across all species; samples with larger petal lengths consistently display proportionally larger petal widths, serving as strong predictive features for classification algorithms.

3. **Balanced Class Distribution:**  
   The dataset features an exact balanced split of **50 samples per species** (150 total), providing an ideal, non-biased benchmark dataset for AI classification models.
